File: scripts/publish-factory.ts
/**
 * FÁBRICA DE POSTS ENEM Pro
 * 
 * Sistema que publica 15-30 posts/dia em horários aleatórios (precisão ms)
 * durante 15 dias. Pega drafts não publicados, roda quality-check,
 * internal-linking, commit, deploy e submete no GSC.
 */

import fs from "fs";
import path from "path";
import { execSync } from "child_process";

const BASE_DIR = "/root/projetos/enem-pro";
const DRAFTS_DIR = path.join(BASE_DIR, ".blog-memory/drafts");
const STATE_FILE = path.join(BASE_DIR, ".blog-memory/publish-queue.json");
const LOG_FILE = path.join(BASE_DIR, ".blog-memory/publish-log.json");

interface PublishState {
  queue: string[];           // drafts waiting to publish
  published: string[];       // already published
  failed: { slug: string; error: string; timestamp: string }[];
  currentBatch: string[];    // posts being published today
  lastRun: string;           // ISO timestamp
  daysRemaining: number;     // days left in 15-day campaign
  totalPublished: number;
}

interface PublishResult {
  success: boolean;
  slug: string;
  error?: string;
  deployUrl?: string;
}

function loadState(): PublishState {
  if (fs.existsSync(STATE_FILE)) {
    try {
      return JSON.parse(fs.readFileSync(STATE_FILE, "utf-8"));
    } catch {
      // fallback
    }
  }
  return {
    queue: [],
    published: [],
    failed: [],
    currentBatch: [],
    lastRun: new Date().toISOString(),
    daysRemaining: 15,
    totalPublished: 0
  };
}

function saveState(state: PublishState): void {
  fs.writeFileSync(STATE_FILE, JSON.stringify(state, null, 2));
}

function log(message: string, level: "info" | "warn" | "error" = "info"): void {
  const timestamp = new Date().toISOString();
  const logEntry = { timestamp, level, message };
  
  let logs: any[] = [];
  if (fs.existsSync(LOG_FILE)) {
    try {
      logs = JSON.parse(fs.readFileSync(LOG_FILE, "utf-8"));
    } catch {}
  }
  logs.push(logEntry);
  // Keep last 1000 entries
  if (logs.length > 1000) logs = logs.slice(-1000);
  fs.writeFileSync(LOG_FILE, JSON.stringify(logs, null, 2));
  
  console.log(`[${timestamp}] ${level.toUpperCase()}: ${message}`);
}

// Get all unpublished drafts
function getUnpublishedDrafts(state: PublishState): string[] {
  const allDrafts = fs.readdirSync(DRAFTS_DIR, { withFileTypes: true })
    .filter(d => d.isDirectory())
    .map(d => d.name);
  
  return allDrafts.filter(slug => 
    !state.published.includes(slug) && 
    !state.failed.some(f => f.slug === slug) &&
    !state.currentBatch.includes(slug) &&
    fs.existsSync(path.join(DRAFTS_DIR, slug, "article.md"))
  );
}

// Pick 15-30 random posts for today
function pickTodaysBatch(state: PublishState): string[] {
  const available = getUnpublishedDrafts(state);
  
  if (available.length === 0) {
    log("Nenhum draft disponível para publicar", "warn");
    return [];
  }
  
  // Random between 15-30
  const count = Math.min(
    Math.floor(Math.random() * (30 - 15 + 1)) + 15,
    available.length,
    30
  );
  
  // Shuffle and pick
  const shuffled = available.sort(() => Math.random() - 0.5);
  const batch = shuffled.slice(0, count);
  
  log(`Batch de hoje: ${batch.length} posts selecionados`);
  return batch;
}

// Convert draft to blog post (uses existing Python script)
function convertDraft(slug: string): boolean {
  try {
    log(`Convertendo draft: ${slug}`);
    execSync("python3 scripts/draft_to_blogpost.py", { 
      cwd: BASE_DIR, 
      stdio: "pipe",
      timeout: 120000
    });
    return true;
  } catch (error: any) {
    log(`Erro ao converter ${slug}: ${error.message}`, "error");
    return false;
  }
}

// Run quality checks
function runQualityCheck(): boolean {
  try {
    log("Rodando quality-check...");
    execSync("node scripts/auto-quality-check.js", { 
      cwd: BASE_DIR, 
      stdio: "pipe",
      timeout: 60000
    });
    return true;
  } catch (error: any) {
    log(`Quality check falhou: ${error.message}`, "error");
    return false;
  }
}

// Run internal linking
function runInternalLinking(): boolean {
  try {
    log("Rodando internal-linking...");
    execSync("node scripts/auto-internal-linking.js", { 
      cwd: BASE_DIR, 
      stdio: "pipe",
      timeout: 60000
    });
    return true;
  } catch (error: any) {
    log(`Internal linking falhou: ${error.message}`, "warn");
    return false; // Non-fatal
  }
}

// Git commit + push
function gitCommitAndPush(slug: string): boolean {
  try {
    log(`Commitando ${slug}...`);
    
    execSync("git add -A", { cwd: BASE_DIR, stdio: "pipe" });
    execSync(`git commit -m "feat: publish ${slug}"`, { cwd: BASE_DIR, stdio: "pipe" });
    execSync("git push origin master", { cwd: BASE_DIR, stdio: "pipe", timeout: 120000 });
    
    log(`Push realizado para ${slug}`);
    return true;
  } catch (error: any) {
    log(`Git push falhou para ${slug}: ${error.message}`, "error");
    return false;
  }
}

// Submit URL to GSC Indexing API
function submitToGSC(slug: string): boolean {
  try {
    const url = `https://questoesenem.pro/blog/${slug}`;
    log(`Submetendo ao GSC: ${url}`);
    
    // Uses the existing submit-sitemap-gsc script
    execSync(`python3 scripts/submit-sitemap-gsc.py "${url}"`, { 
      cwd: BASE_DIR, 
      stdio: "pipe",
      timeout: 30000
    });
    
    return true;
  } catch (error: any) {
    log(`GSC submit falhou para ${slug}: ${error.message}`, "warn");
    return false; // Non-fatal
  }
}

// Publish a single post
async function publishPost(slug: string): Promise<PublishResult> {
  log(`=== INICIANDO PUBLICAÇÃO: ${slug} ===`);
  
  try {
    // 1. Convert draft
    if (!convertDraft(slug)) {
      return { success: false, slug, error: "Falha na conversão do draft" };
    }
    
    // 2. Quality check
    if (!runQualityCheck()) {
      return { success: false, slug, error: "Falha no quality-check" };
    }
    
    // 3. Internal linking
    runInternalLinking(); // Non-blocking
    
    // 4. Git commit + push (triggers Vercel deploy)
    if (!gitCommitAndPush(slug)) {
      return { success: false, slug, error: "Falha no git push" };
    }
    
    // 5. Wait a bit for Vercel deploy
    log("Aguardando deploy Vercel (30s)...");
    await new Promise(r => setTimeout(r, 30000));
    
    // 6. Submit to GSC
    submitToGSC(slug);
    
    log(`=== PUBLICADO COM SUCESSO: ${slug} ===`);
    return { success: true, slug };
    
  } catch (error: any) {
    log(`Erro inesperado em ${slug}: ${error.message}`, "error");
    return { success: false, slug, error: error.message };
  }
}

// Main daily runner
async function runDailyBatch() {
  log("=== INICIANDO BATCH DIÁRIO ===");
  
  const state = loadState();
  
  // Check if campaign is done
  if (state.daysRemaining <= 0) {
    log("Campanha de 15 dias concluída!", "info");
    return;
  }
  
  // Pick today's batch
  const batch = pickTodaysBatch(state);
  
  if (batch.length === 0) {
    log("Nenhum post para publicar hoje", "warn");
    return;
  }
  
  state.currentBatch = batch;
  saveState(state);
  
  // Schedule each post at random time today (ms precision)
  const now = new Date();
  const endOfDay = new Date(now);
  endOfDay.setHours(23, 59, 59, 999);
  
  for (const slug of batch) {
    // Random time between now and end of day
    const randomMs = Math.floor(Math.random() * (endOfDay.getTime() - now.getTime()));
    const scheduledTime = new Date(now.getTime() + randomMs);
    
    log(`Agendado ${slug} para ${scheduledTime.toISOString()}`);
    
    // In production, this would be a separate cron job
    // For now, we publish sequentially with small delays
    await new Promise(r => setTimeout(r, 1000)); // 1s between posts
    
    const result = await publishPost(slug);
    
    if (result.success) {
      state.published.push(slug);
      state.totalPublished++;
    } else {
      state.failed.push({
        slug,
        error: result.error || "Erro desconhecido",
        timestamp: new Date().toISOString()
      });
    }
    
    // Remove from current batch
    state.currentBatch = state.currentBatch.filter(s => s !== slug);
    saveState(state);
  }
  
  // Update campaign progress
  state.daysRemaining--;
  state.lastRun = new Date().toISOString();
  saveState(state);
  
  log(`=== BATCH DIÁRIO CONCLUÍDO ===`);
  log(`Publicados hoje: ${batch.filter(s => state.published.includes(s)).length}/${batch.length}`);
  log(`Total publicado: ${state.totalPublished}`);
  log(`Dias restantes: ${state.daysRemaining}`);
}

// Run if called directly
if (import.meta.url === `file://${process.argv[1]}`) {
  runDailyBatch().catch(err => {
    log(`Erro fatal: ${err.message}`, "error");
    process.exit(1);
  });
}

export { runDailyBatch, loadState, saveState };

File: scripts/publish_factory.py
#!/usr/bin/env python3
"""
FÁBRICA DE POSTS ENEM Pro - Versão Python
Publica 15-30 posts/dia em horários aleatórios (precisão ms) durante 15 dias.
"""

import os
import json
import random
import subprocess
import time
from datetime import datetime, timedelta
from pathlib import Path

BASE_DIR = Path("/root/projetos/enem-pro")
DRAFTS_DIR = BASE_DIR / ".blog-memory" / "drafts"
STATE_FILE = BASE_DIR / ".blog-memory" / "publish-queue.json"
LOG_FILE = BASE_DIR / ".blog-memory" / "publish-log.json"

def load_state():
    if STATE_FILE.exists():
        try:
            return json.loads(STATE_FILE.read_text())
        except:
            pass
    return {
        "queue": [],
        "published": [],
        "failed": [],
        "current_batch": [],
        "last_run": datetime.now().isoformat(),
        "days_remaining": 15,
        "total_published": 0
    }

def save_state(state):
    STATE_FILE.write_text(json.dumps(state, indent=2, ensure_ascii=False))

def log(message, level="info"):
    timestamp = datetime.now().isoformat()
    entry = {"timestamp": timestamp, "level": level, "message": message}
    
    logs = []
    if LOG_FILE.exists():
        try:
            logs = json.loads(LOG_FILE.read_text())
        except:
            pass
    logs.append(entry)
    if len(logs) > 1000:
        logs = logs[-1000:]
    LOG_FILE.write_text(json.dumps(logs, indent=2, ensure_ascii=False))
    
    print(f"[{timestamp}] {level.upper()}: {message}")

def get_unpublished_drafts(state):
    all_drafts = [d.name for d in DRAFTS_DIR.iterdir() if d.is_dir()]
    return [s for s in all_drafts 
            if s not in state["published"] 
            and s not in [f["slug"] for f in state["failed"]]
            and s not in state["current_batch"]
            and (DRAFTS_DIR / s / "article.md").exists()]

def pick_todays_batch(state):
    available = get_unpublished_drafts(state)
    if not available:
        log("Nenhum draft disponível", "warn")
        return []
    
    count = min(random.randint(15, 30), len(available), 30)
    batch = random.sample(available, count)
    log(f"Batch de hoje: {len(batch)} posts")
    return batch

def run_cmd(cmd, cwd=BASE_DIR, timeout=120):
    try:
        result = subprocess.run(cmd, shell=True, cwd=cwd, capture_output=True, text=True, timeout=timeout)
        return result.returncode == 0, result.stdout, result.stderr
    except subprocess.TimeoutExpired:
        return False, "", "Timeout"
    except Exception as e:
        return False, "", str(e)

def convert_draft(slug):
    log(f"Convertendo: {slug}")
    ok, out, err = run_cmd("python3 scripts/draft_to_blogpost.py", timeout=120)
    return ok

def run_quality_check():
    log("Quality check...")
    ok, out, err = run_cmd("node scripts/auto-quality-check.js", timeout=60)
    return ok

def run_internal_linking():
    log("Internal linking...")
    ok, out, err = run_cmd("node scripts/auto-internal-linking.js", timeout=60)
    return ok

def git_commit_push(slug):
    log(f"Git commit+push: {slug}")
    ok, out, err = run_cmd("git add -A", timeout=30)
    if not ok:
        return False
    ok, out, err = run_cmd(f'git commit -m "feat: publish {slug}"', timeout=30)
    if not ok:
        return False
    ok, out, err = run_cmd("git push origin master", timeout=120)
    return ok

def submit_gsc(slug):
    log(f"GSC submit: {slug}")
    url = f"https://questoesenem.pro/blog/{slug}"
    ok, out, err = run_cmd(f'python3 scripts/submit-sitemap-gsc.py "{url}"', timeout=30)
    return ok

def publish_post(slug):
    log(f"=== PUBLICANDO: {slug} ===")
    
    try:
        # 1. Convert
        if not convert_draft(slug):
            return False, "Falha conversão"
        
        # 2. Quality check
        if not run_quality_check():
            return False, "Falha quality-check"
        
        # 3. Internal linking (non-blocking)
        run_internal_linking()
        
        # 4. Git commit + push
        if not git_commit_push(slug):
            return False, "Falha git push"
        
        # 5. Wait for Vercel deploy
        log("Aguardando Vercel deploy (30s)...")
        time.sleep(30)
        
        # 6. GSC submit
        submit_gsc(slug)
        
        log(f"=== SUCESSO: {slug} ===")
        return True, None
        
    except Exception as e:
        log(f"Erro em {slug}: {e}", "error")
        return False, str(e)

def run_daily_batch():
    log("=== BATCH DIÁRIO INICIADO ===")
    state = load_state()
    
    if state["days_remaining"] <= 0:
        log("Campanha de 15 dias concluída!")
        return
    
    batch = pick_todays_batch(state)
    if not batch:
        return
    
    state["current_batch"] = batch
    save_state(state)
    
    # Schedule random times for today
    now = datetime.now()
    end_of_day = now.replace(hour=23, minute=59, second=59, microsecond=999999)
    total_seconds = (end_of_day - now).total_seconds()
    
    for i, slug in enumerate(batch):
        # Random delay up to end of day
        delay = random.uniform(0, total_seconds)
        scheduled = now + timedelta(seconds=delay)
        log(f"Agendado {slug} para {scheduled.isoformat()}")
        
        # For now, run sequentially with small delay
        if i > 0:
            time.sleep(random.uniform(1, 5))  # 1-5s between posts
        
        success, error = publish_post(slug)
        
        if success:
            state["published"].append(slug)
            state["total_published"] += 1
        else:
            state["failed"].append({
                "slug": slug,
                "error": error,
                "timestamp": datetime.now().isoformat()
            })
        
        state["current_batch"] = [s for s in state["current_batch"] if s != slug]
        save_state(state)
    
    state["days_remaining"] -= 1
    state["last_run"] = datetime.now().isoformat()
    save_state(state)
    
    log("=== BATCH DIÁRIO CONCLUÍDO ===")
    today_published = len([s for s in batch if s in state["published"]])
    log(f"Publicados hoje: {today_published}/{len(batch)}")
    log(f"Total: {state['total_published']}")
    log(f"Dias restantes: {state['days_remaining']}")

if __name__ == "__main__":
    run_daily_batch()